Add Rust Candidate block

boydjohnson opened the pull request:
This is built off of #1673.

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -0,0 +1,107 @@
+    ) -> Self {
+        let bs = unsafe {
+            let py = cpython::Python::assume_gil_acquired();
+            block_store.clone_ref(py)
```
Why are we cloning this refs?

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
+                )
+                .expect("consensus has no method 'check_publish_block'")
+                .extract::<bool>(py)
+                .unwrap()
```
put this in a function please

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
+            .collect();
+
+        let builder = {
+            let py = unsafe { cpython::Python::assume_gil_acquired() };
```
We should try to only call this once at the beginning of the function if it is needed because the GIL is already in scope for the entire function.

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
+        let mut bad_batches = vec![];
+        let mut pending_batches = vec![];
+
+        for batch in self.pending_batches.clone() {
```
I think you can probably get rid of this clone, either by doing `.iter()` or using `std::mem::replace`

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -220,4 +227,180 @@ impl CandidateBlock {
+                    bad_batches.push(batch.clone());
+                    pending_batches.clear();
+                    pending_batches.append(&mut self.pending_batches
+                        .clone()
```
If you `std::mem::replace` this probably becomes easier too (doesn't require clone())

aludvik commented on validator/src/journal/candidate_block.rs:
```diff
@@ -162,4 +162,47 @@ impl CandidateBlock {
+
+            // TODO: call Poll Injector python functions
+
+            // TODO: Enforce validation rules
```
Should these be done?

boydjohnson commented:
@aludvik I implemented the 2 todos in add_batch, removed the spurious clone_ref, and reduced the number of calls of `.assume_gil_acquired`. I am not sure how to reduce the clones on `self.pending_batches`. I think it is either clone every time we add a batch to the the new `pending_batches` vec or clone `self.pending_batches` to begin with. I have wondered if I should spend some time making the signature into `finalize(self)` instead of `finalize(&mut self)` since it should consume the candidate block.
